DYDX-Bots/Large-Inactive-Balance/src/agent.ts

```diff
@@ -0,0 +1,61 @@
+import { BigNumber, providers } from "ethers";
+import { Finding, getEthersProvider, HandleTransaction, TransactionEvent } from "forta-agent";
+import BalanceFetcher from "./balance.fetcher";
+import InactiveBalanceFetcher from "./inactive.balance.fetcher";
+import { createFinding, EVENT_ABI } from "./utils";
+import { BotConfig, STATIC_CONFIG, DYNAMIC_CONFIG } from "./config";
+import NetworkData, { NETWORK_MAP } from "./network";
+import NetworkManager from "./network";
+
+const networkManager = new NetworkManager(NETWORK_MAP);
+const balanceFetcher = new BalanceFetcher(getEthersProvider(), networkManager);
+const inactiveBalanceFetcher = new InactiveBalanceFetcher(getEthersProvider(), networkManager);
+
+export const provideInitialize = (provider: providers.Provider) => async () => {
+  const { chainId } = await provider.getNetwork();
+  networkManager.setNetwork(chainId);
+  balanceFetcher.setTokenContract();
+  inactiveBalanceFetcher.setSafetyModule();
+};
+
+export const provideHandleTransaction =
+  (
+    config: BotConfig,
+    networkManager: NetworkData,
+    inactiveBalanceFetcher: InactiveBalanceFetcher,
+    balanceFetcher: BalanceFetcher
+  ): HandleTransaction =>
+  async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    await Promise.all(
+      txEvent.filterLog(EVENT_ABI, networkManager.safetyModule).map(async (log) => {
+        // get the staker address
+        const staker = log.args.staker;
+
+        // get the staker inactive balance.
+        const inactiveBalance = await inactiveBalanceFetcher.getInactiveBalance(staker, txEvent.blockNumber);
+
+        // set threshold based on mode.
+        let _threshold: BigNumber;
+
+        if (config.mode === "STATIC") _threshold = config.thresholdData;
+        else {
+          // fetch total staked tokens.
+          const totalStaked = await balanceFetcher.getBalance(txEvent.blockNumber);
```

Comment:
Not necessary to fetch the balance at the previous block?

Reply:
I believe it is not necessary in this bot since we are comparing two values at the end of the same block. Nothing is compared with the event args so there should be no incoherence.